lib/pacemaker/pcmk_sched_probes.c

```diff
@@ -503,28 +503,37 @@ pcmk__order_probes(pe_working_set_t *data_set)
     order_then_probes(data_set);
 }
 
-/*
- * Check nodes for resources started outside of the LRM
+/*!
+ * \internal
+ * \brief Schedule any probes needed
+ *
+ * \param[in] data_set  Cluster working set
+ *
+ * \note This may also schedule fencing of failed remote nodes.
+ * \todo This could benefit from optimization. A test in 2010 using 100 nodes
+ *       and 100 clones completed scheduling in 4 seconds without probes, and
+ *       42 seconds with probes. The test should be repeated and profiled.
```

Comment:
Hm, I wonder how that test was performed (aside from the very brief comment about "ptest -x").  I don't see anything in our existing test suites that allows for checking timing data.  Maybe it would be worth adding a scheduler test for this and making it part of the standard tests.  Then we'll know if it's still slow after all these years, and then we can work on speeding it up.

Reply:
I've been playing around with this a bit and have yet to reproduce the slowness, though maybe I've misconfigured my input file for crm_simulate.  It looks like this (with tons and tons of lines removed):
```
<cib crm_feature_set="3.3.0" validate-with="pacemaker-3.3" epoch="1" num_updates="1342" admin_epoch="1" cib-last-written="Tue Jan  4 10:00:00 2021" update-origin="cluster01" update-client="crmd" update-user="hacluster" have-quorum="1" dc-uuid="2">
  <configuration>
    <crm_config>
      <cluster_property_set id="cib-bootstrap-options">
        <nvpair id="cib-bootstrap-options-have-watchdog" name="have-watchdog" value="false"/>
        <nvpair id="cib-bootstrap-options-dc-version" name="dc-version" value="2.0.4-1.e97f9675f.git.el7-e97f9675f"/>
        <nvpair id="cib-bootstrap-options-cluster-infrastructure" name="cluster-infrastructure" value="corosync"/>
        <nvpair id="cib-bootstrap-options-cluster-name" name="cluster-name" value="test-cluster"/>
        <nvpair id="cib-bootstrap-options-stonith-enabled" name="stonith-enabled" value="true"/>
        <nvpair id="cib-bootstrap-options-maintenance-mode" name="maintenance-mode" value="false"/>
      </cluster_property_set>
    </crm_config>
    <nodes>
      <node id="001" uname="cluster001"/>
...
      <node id="100" uname="cluster100"/>
    </nodes>
    <resources>
      <primitive class="stonith" id="Fencing" type="fence_xvm">
        <instance_attributes id="Fencing-instance_attributes">
          <nvpair id="Fencing-instance_attributes-ip_family" name="ip_family" value="ipv4"/>
        </instance_attributes>
        <operations>
          <op id="Fencing-monitor-interval-60s" interval="60s" name="monitor"/>
        </operations>
      </primitive>
      <clone id="dummy-clone">
        <primitive class="ocf" id="dummy" provider="pacemaker" type="Dummy"/>
        <meta_attributes id="dummy-clone.meta">
          <nvpair id="nvpair.meta.clone_max" name="clone-max" value="100"/>
        </meta_attributes>
      </clone>
    </resources>
    <constraints/>
  </configuration>
  <status>
    <node_state id="001" uname="cluster001" in_ccm="true" crmd="online" join="member" expected="member" crm-debug-origin="crm_simulate">
      <lrm id="001">
        <lrm_resources>
          <lrm_resource id="Fencing" class="stonith" type="fence_xvm">
            <lrm_rsc_op id="Fencing_last_0" operation_key="Fencing_monitor_0" operation="monitor" crm-debug-origin="crm_simulate" crm_feature_set="3.11.0" transition-key="2:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" transition-magic="0:0;2:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" exit-reason="" call-id="2" rc-code="0" op-status="0" interval="0" last-rc-change="1641323443" exec-time="0" queue-time="0" op-digest="7da16842ab2328e41f737cab5e5fc89c"/>
            <lrm_rsc_op id="Fencing_monitor_60000" operation_key="Fencing_monitor_60000" operation="monitor" crm-debug-origin="crm_simulate" crm_feature_set="3.11.0" transition-key="3:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" transition-magic="0:0;3:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" exit-reason="" call-id="3" rc-code="0" op-status="0" interval="60000" last-rc-change="1641323443" exec-time="0" queue-time="0" op-digest="a88218bb6c7dc47e6586fc75fc2a8d69"/>
          </lrm_resource>
          <lrm_resource id="dummy" class="ocf" provider="pacemaker" type="Dummy">
            <lrm_rsc_op id="dummy_last_0" operation_key="dummy_monitor_0" operation="monitor" crm-debug-origin="crm_simulate" crm_feature_set="3.11.0" transition-key="2:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" transition-magic="0:0;2:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" exit-reason="" call-id="2" rc-code="7" op-status="4" interval="0" last-rc-change="1641323443" exec-time="0" queue-time="0" op-digest="f2317cad3d54cec5d7d7aa7d0bf35cf8"/>
          </lrm_resource>
        </lrm_resources>
      </lrm>
    </node_state>
...
    <node_state id="100" uname="cluster100" in_ccm="true" crmd="online" join="member" expected="member" crm-debug-origin="crm_simulate">
      <lrm id="100">
        <lrm_resources>
          <lrm_resource id="Fencing" class="stonith" type="fence_xvm">
            <lrm_rsc_op id="Fencing_last_0" operation_key="Fencing_monitor_0" operation="monitor" crm-debug-origin="crm_simulate" crm_feature_set="3.11.0" transition-key="1:-1:7:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" transition-magic="0:7;1:-1:7:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" exit-reason="" call-id="1" rc-code="7" op-status="0" interval="0" last-rc-change="1641323467" exec-time="0" queue-time="0" op-digest="7da16842ab2328e41f737cab5e5fc89c"/>
          </lrm_resource>
          <lrm_resource id="dummy" class="ocf" provider="pacemaker" type="Dummy">
            <lrm_rsc_op id="dummy_last_0" operation_key="dummy_monitor_0" operation="monitor" crm-debug-origin="crm_simulate" crm_feature_set="3.11.0" transition-key="2:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" transition-magic="0:0;2:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" exit-reason="" call-id="2" rc-code="7" op-status="4" interval="0" last-rc-change="1641323467" exec-time="0" queue-time="0" op-digest="f2317cad3d54cec5d7d7aa7d0bf35cf8"/>
          </lrm_resource>
        </lrm_resources>
      </lrm>
    </node_state>
  </status>
</cib>
```

crm_simulate finishes pretty much instantaneously.

Reply:
ptest was what eventually became crm_simulate.

The issue isn't parsing probe results, it's scheduling the probes themselves -- i.e. your node_state entries should be empty so probes are scheduled. It may also be interesting to compare against a unique clone, i.e. probes will have to be scheduled separately for every instance of the clone.

If there's no obvious time difference, passing --repeat 1000 might show something.

Reply:
With all the node_state entries removed, I see:
```
clumens@spire ~/src/pacemaker master|✔❯ tools/crm_simulate -S --profile ~/tst                
Timings:
  * Testing /home/clumens/tst/huge.xml ... 0.18 secs
clumens@spire ~/src/pacemaker master|✔❯ tools/crm_simulate -S --profile ~/tst -N 1000
Timings:
  * Testing /home/clumens/tst/huge.xml ... 171.85 secs
```
I have noy yet gotten around to testing the unique clone suggestion.

Reply:
Cool. The main question would be how the times compare to commenting out the call to pcmk__schedule_probes() in pcmk_sched_allocate.c.

Reply:
If I comment out the probe_resources block in stage5 in my tree, there is no significant difference:
```
clumens@spire ~/src/pacemaker master↑6|✔❯ tools/crm_simulate -S --profile ~/tst -N 1000
Timings:
  * Testing /home/clumens/tst/huge.xml ... 170.15 secs
```

Reply:
Awesome, I'll delete the comment :)

Reply:
added commit on top